### ISSUE TITLE
Implement BMI1 and BMI2 intrinsics

### ISSUE THREAD

**Issue by tannergooding** [dotnet/coreclr#18712]
Pending BMI1:
- [x] AndNot
- [ ] BitFieldExtract
- [x] ExtractLowestSetBit
- [x] GetMaskUptoLowestSetBit
- [x] ResetLowestSetBit
- [x] TrailingZeroCount

Pending BMI2:
- [ ] MultiplyNoFlags
- [x] ParallelBitDeposit
- [x] ParallelBitExtract
- [ ] ZeroHighBits

**Comment by tannergooding** [comment]
Logged issue and self-assigned. I have most of this done locally.

**Comment by tannergooding** [comment]
Some of the type parameters for `BitFieldExtract` should be changed. Specifically, `start` and `length` should both be `bytes`.

`GetMaskUptoLowestSetBit` should be renamed to correct the casing.

**Comment by tannergooding** [comment]
FYI. @eerhardt, @fiigii, @CarolEidt

**Comment by mikedn** [comment]
`AndNot` is pointless as intrinsics, it should be easy to generate automatically. And `MultiplyNoFlags` is probably pointless since "no flags" is not something that has meaning in C#.

**Comment by GrabYourPitchforks** [comment]
`pext` and `pdep` in particular would be useful for the UTF-8 processing work we're doing.

**Comment by voinokin** [comment]
My 5c
ANDN can be used for branchless select operation in absence of CMOV with the following sequence:
- eval cc
- SETcc sel + zero-extend
- NEG sel
- AND r1, sel
- ANDN r2, sel
- OR r1, r2
Not the fastest solution, still saves 1 NOT op and potentially 1 reg wrt to just ANDs

**Comment by fiigii** [comment]
> Specifically, start and length should both be bytes.

@tannergooding Thank you for the work. Could you elaborate on this?

**Comment by tannergooding** [comment]
@fiigii, those values (start and length) are combined and encoded in operand 3 (VEX.vvvv) of the instruction.

> Bit 7:0 of the second source operand specifies the starting bit position of bit extraction. A START value exceeding the operand size will not extract any bits from the second source operand. Bit 15:8 of the second source operand specifies the maximum number of bits (LENGTH) beginning at the START position to extract.

> BEXTR r32a, r/m32, r32b

**Comment by fiigii** [comment]
Yes, but these arguments will be promoted to 32bit in the runtime even if they are byte.

**Comment by tannergooding** [comment]
> Yes, but these arguments will be promoted to 32bit in the runtime even if they are byte.

They are currently set to `uint` for the 32-bit overloads and `ulong` for the 64-bit overloads. At the very least, they should be `uint` on both overloads.

I had suggested changing to `byte`, as that better represents the actual value range and how we have encoded similar operands on other instrinsic methods.

**Comment by fiigii** [comment]
Agree, thanks for the explanation.

**Comment by tannergooding** [comment]
This is mostly just waiting on the API fixes to flow back from CoreFX so that I can finish validating the implementation with tests.

**Comment by tannergooding** [comment]
@GrabYourPitchforks, the PR implementing `pdep` and `pext` is up here: https://github.com/dotnet/coreclr/pull/18785

It will probably be sometime next week before it is actually merged, due to the Jenkins outages.

**Comment by pentp** [comment]
MULX is definitely useful (at least the 64x64->128 variant).
From https://github.com/dotnet/coreclr/issues/17818#issuecomment-395365269: `MultiplyNoFlags` should have an overload with only the high half returned (this saves one register) - `ulong MultiplyNoFlagsHigh(ulong left, ulong right)`

**Comment by tannergooding** [comment]
> Is there a chance you could describe what I should put in global.json try out BMI1+2?

@damageboy, you need to wait for a public preview containing `.NET Core 3.0` or you need to build the bits yourself (locally).

> It seems to me that the nupkg for the intrinsics is still not ready? Or am I looking at the wrong place?

The intrinsics package only contained the API surface area (not the implementation, which requires CoreCLR). It was also only for previewing the bits in 2.1.
For 3.0, they will be in box and no nuget package will be required (or even available) to use the APIs (they will just be available if you reference `netcoreapp3.0`).

**Comment by eerhardt** [comment]
>you need to wait for a public preview containing .NET Core 3.0 or you need to build the bits yourself (locally).

You can also install a nightly build of `.NET Core 3.0` using the "master" branch/column here: https://github.com/dotnet/core-sdk#installers-and-binaries

**Comment by damageboy** [comment]
Got it, thanks for the nupkg clearup.

Is 3.0 (master) is a consumable form right now?
e.g. will it be an adventure or a suicide attempt?
I'm ok with compiling like the wind.

**Comment by damageboy** [comment]
My next question, as I'm working a real-life use-case of intrinsics, is would this be a good time to report on CQ issues if/when those arise?

**Comment by eerhardt** [comment]
3.0 (master) is consumable, we try to keep our branches in a clean state at all times (at least tests passing). You may not get the absolute latest runtime code when you install the SDK, but it shouldn't be too old (it looks like the latest version is currently [2 days old](https://github.com/dotnet/core-sdk/blob/82bd85d0a9672b5f80f1a2d9802583a5e53b2d6d/build/DependencyVersions.props#L20).

If you are just doing investigation/research, a "trick" I like to do is to download the SDK's .zip (not the `.exe` installer), extract is somewhere temporary (like `D:\dotnet`) and then put that on my `%PATH%` just for a single command prompt. If you launch VS from that command prompt, VS will use the temporary SDK, since it is the first one on the `%PATH%`.
Then when you are done, you can just close the command prompt, delete `D:\dotnet`, and your machine is left in a 100% clean state.

**Comment by damageboy** [comment]
@eerhardt Thanks, sounds like a plan.
I'm doing this on linux, but hopefully my command line experience will stay the same,
Are there any tips for vscode / rider users?

**Comment by CarolEidt** [comment]
> My next question, as I'm working a real-life use-case of intrinsics, is would this be a good time to report on CQ issues if/when those arise?

Yes, that would be great. There are already some known issues with CQ for vector types specifically and for structs generally. You can search for issues in coreclr with the label "tenet-performance" - and usually "area-CodeGen". They will generally have "struct" or "SIMD" or "vector" in their title.

**Comment by eerhardt** [comment]
> Are there any tips for vscode / rider users?

Unfortunately I've never used rider.  But I believe vscode uses the first SDK on the `$PATH` as well.

If you happen to be using VS for Mac, they have a configuration setting for where your "dotnet hive" is:

See the `Allow .NET Core location to be configured` section in https://lastexitcode.com/blog/2018/05/20/NetCoreSupportInVisualStudioMac7-5/.

**Comment by damageboy** [comment]
Just wanted to update that things are looking very good as far as my tests are concerned.

I'm trying to coerce BDN to run some benchmarks, and I will try to get disassembly out of the JIT, as I do have C++ versions of some functions I can directly compare to.

**Comment by damageboy** [comment]
If I understand correctly, the only reasonable way to get disassembly right now is to build coreclr locally and try following [this](https://github.com/dotnet/coreclr/blob/master/Documentation/building/viewing-jit-dumps.md), right?

There's no magic thing that I can pull off with a local .tar.gz download of sdk 3.0 downloaded from [here](https://github.com/dotnet/core-sdk#installers-and-binaries) right?

**Comment by AndyAyersMS** [comment]
Benchmark.Net has a [disassembly diagnoser](http://adamsitnik.com/Disassembly-Diagnoser/) which doesn't require special builds. You can also get at the jit-generated code via the debugger.

**Comment by damageboy** [comment]
@AndyAyersMS Thanks, I'm aware of it, but still having to fight with BDN on the custom toolchain (encountering weird error)... Also I'm on Linux, so IIRC the diagnoser would not be available even I had  a working BDN setup...

**Comment by damageboy** [comment]
Hi everyone, 
I've managed to finish my test of Popcnt, BMI1, BMI2 intrinsics and I must say it was a very pleasant experience.

I managed to inspect the generated code with my own locally compiled CoreCLR on linux

I'm pretty sure I am hitting up a relatively common Intel CPU errata when it comes to popcnt / tzcnt / lzcnt and friends:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=62011

I suspect that the only solution to this is to insert seemingly needless `xor dest,dest` instructions into the generated loop code to avoid the false dependency described in the gcc issue.

I assume you would want a separate issue with JIT dumps on this?

**Comment by damageboy** [comment]
Hi, I've opened #19555 with what I think is a CQ issue with false dependency I mentioned above

**Comment by AndyAyersMS** [comment]
Thanks for following up on this.

**Comment by damageboy** [comment]
On possibly related news, I'm seeing something weird when looking in the JitDump of the following function:

```csharp
public static unsafe int Intrinsics(ulong* bits, int numBits, int n)
{
    var p64 = bits;
    var prevN = n;
    while (n > 0) {
        prevN = n;
        n -= (int) PopCount(*(p64++));
    }
    var pos = (int) TrailingZeroCount(ParallelBitDeposit(1UL << (prevN - 1), *(p64 - 1)));
    return (int) ((p64 - 1 - bits) << 6) + pos;
}
```

When looking in the JITDump, generated on .NET Core 3.0, I'm seeing this "pattern" (below):
```nasm
G_M5400_IG01:        ; func=00, offs=000000H, size=000AH, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
IN0025: 000000 push     rbp
IN0026: 000001 sub      rsp, 16
IN0027: 000005 lea      rbp, [rsp+10H]
G_M5400_IG02:        ; offs=00000AH, size=0018H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, isz
IN0001: 00000A mov      rcx, 0x7F37BFDE4498
IN0002: 000014 mov      rax, qword ptr [rcx]
IN0003: 000017 mov      ecx, edi
IN0004: 000019 mov      rdi, rax
IN0005: 00001C mov      esi, ecx
IN0006: 00001E test     esi, esi
IN0007: 000020 jle      SHORT G_M5400_IG07
G_M5400_IG03:        ; offs=000022H, size=0019H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, isz
IN0008: 000022 mov      esi, ecx
IN0009: 000024 lea      rcx, [rdi+8]
IN000a: 000028 mov      qword ptr [V03 rbp-08H], rcx
IN000b: 00002C popcnt   rdi, qword ptr [rdi]
IN000c: 000031 mov      edx, esi
IN000d: 000033 sub      edx, edi
IN000e: 000035 test     edx, edx
IN000f: 000037 mov      ecx, edx
IN0010: 000039 jg       SHORT G_M5400_IG06
G_M5400_IG04:        ; offs=00003BH, size=003BH, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref
IN0011: 00003B lea      ecx, [rsi-1]
IN0012: 00003E mov      edi, 1
IN0013: 000043 shl      rdi, cl
IN0014: 000046 mov      rcx, qword ptr [V03 rbp-08H]
IN0015: 00004A vpdep    rdi, rdi, qword ptr [rcx-8]
IN0016: 000050 tzcnt    rdi, rdi
IN0017: 000055 add      rcx, -8
IN0018: 000059 mov      rsi, rcx
IN0019: 00005C sub      rsi, rax
IN001a: 00005F mov      rax, rsi
IN001b: 000062 sar      rax, 63
IN001c: 000066 and      rax, 7
IN001d: 00006A add      rax, rsi
IN001e: 00006D sar      rax, 3
IN001f: 000071 shl      eax, 6
IN0020: 000074 add      eax, edi
G_M5400_IG05:        ; offs=000076H, size=0006H, epilog, nogc, emitadd
IN0028: 000076 lea      rsp, [rbp]
IN0029: 00007A pop      rbp
IN002a: 00007B ret
G_M5400_IG06:        ; offs=00007CH, size=0006H, gcVars=0000000000000000 {}, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, gcvars, byref, isz
IN0021: 00007C mov      rdi, qword ptr [V03 rbp-08H]
IN0022: 000080 jmp      SHORT G_M5400_IG03
G_M5400_IG07:        ; offs=000082H, size=0006H, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, isz
IN0023: 000082 mov      qword ptr [V03 rbp-08H], rdi
IN0024: 000086 jmp      SHORT G_M5400_IG04
```
specifically, I'm referring to the part where the loop @ `G_M5400_IG03` jumps, near its end, to `G_M5400_IG06` just to jump back to `G_M5400_IG03` again.

Now, granted, this might be the smartest piece of code generation and just me being dumb (highly likely), but I'm investigating a perf regression w.r.t to this `POPCNT` loop, where the same code in certain variants of this function runs considerably faster (25% faster!) depending on what comes AFTER the tight `POPCNT` loop which dominated 95% of the function runtime anyway (yes, I just wrote that).

I was trying to conceive a way that this could be cause by some wacky branch-prediction-aliasing, but from what I know, the jmp instructions at the end of the blocks `IG06`, `IG07`, being unconditional, should NOT be affected by the branch predictor anyway. So I don't think that theory has any support.

Other than that, I'm out of ideas...

Is it possible that this JIT behavior could account for so much perf jitter/loss/gain?

If this should be a separate issue and I'm just littering this BMI1/BMI2 issue, let me know... I'll be happy to open this elsewhere and discuss there...

**Comment by damageboy** [comment]
@AndyAyersMS You are definitely right about the reload being the difference between the good/bad version.

I've uploaded two d1.txt, d2.txt where:
* [d1.txt](https://github.com/dotnet/coreclr/files/2304202/d1.txt) is the JITDump for the following inlined function:
```csharp
public static unsafe int Intrinsics(ulong* bits, int numBits, int n)
{
    var p64 = bits;
    var prevN = n;
    while (n > 0) {
        prevN = n;
        n -= (int) PopCount(*(p64++));
    }
    var pos = (int) TrailingZeroCount(ParallelBitDeposit(1UL << (prevN - 1), *(p64 - 1)));
    return (int) ((p64 - 1 - bits) << 6) + pos;
}
```
* [d2.txt](https://github.com/dotnet/coreclr/files/2304203/d2.txt) is the JITDump for this close variant:
```csharp
public static unsafe int POPCNTButNoBMI2(ulong* bits, int numBits, int n)
{
    var p64 = bits;

    var prevN = n;
    while (n > 0) {
        prevN = n;
        n -= (int) PopCount(*(p64++));
    }

    var p32 = (uint *) (p64 - 1);
    n = prevN - PopCount(*p32);
    if (n > 0) {
        prevN = n;
        p32++;
    }

    var prevValue = *p32;
    var pos = (p32 - (uint *) bits) * 32;
    while (prevN > 0) {
        var bp = TrailingZeroCount(prevValue) + 1;
        pos += bp;
        prevN--;
        prevValue >>= (int) bp;
    }

    return (int) pos - 1;
}
```
Where d1 runs 33% slower than d2, and the biggest noticeable change is the reload that is happenning outside the main loop body.

I couldn't find any of the "cannot be enregistered" notes in the dump, so I uploaded both of them...

**Comment by AndyAyersMS** [comment]
If you get rid of the inline postincrement, things get a bit simpler for the allocator.
```C#
public static unsafe int Intrinsics(ulong* bits, int numBits, int n)
{
    var p64 = bits;
    var prevN = n;
    while (n > 0) {
        prevN = n;
        n -= (int) Popcnt.PopCount(*p64);
        p64++;
    }
    var pos = (int) Bmi1.TrailingZeroCount(
        Bmi2.ParallelBitDeposit(1UL << (prevN - 1), *(p64 - 1)));
    return (int) ((p64 - 1 - bits) << 6) + pos;
}
```
Latest master gives a contiguous inner loop:
```asm
G_M40322_IG03:
       418BC8               mov      ecx, r8d
       F34C0FB802           popcnt   r8, qword ptr [rdx]
       448BC9               mov      r9d, ecx
       452BC8               sub      r9d, r8d
       458BC1               mov      r8d, r9d
       4883C208             add      rdx, 8
       4585C0               test     r8d, r8d
       7FE6                 jg       SHORT G_M40322_IG03
```

**Comment by damageboy** [comment]
Thanks! That certainly did part of the job, I started playing more and more with the code, and eventually managed to improve the runtime quite drastically...

It's very touchy though...

**Comment by CarolEidt** [comment]
The splitting of back-edges (and the undesirability of splitting edges in general) is a known issue with the register allocator: https://github.com/dotnet/coreclr/issues/16857.

This is, IMO, one of the higher priority register allocation issues, and I hope to have time to address it in the near future.

**Comment by damageboy** [comment]
I posted something on my blog about my experience of doing BMI1+2 with CoreCLR 3.0:

https://bits.houmus.org/2018-08-18/netcoreapp3.0-instrinsics-in-real-life-pt1

https://bits.houmus.org/2018-08-19/netcoreapp3.0-instrinsics-in-real-life-pt2

**Comment by tannergooding** [comment]
@fiigii, looks like we still have 3 APIs that aren't implemented yet.

Is this something you can look at while I finish up the helper intrinsics?

**Comment by pentp** [comment]
Is it too late to add an overload of MULX that only returns the high half (`ulong MultiplyNoFlagsHigh(ulong left, ulong right)`)? This can be emitted so that only one register is overwritten. Or maybe the JIT can automatically detect that the lower half isn't used?

**Comment by mikedn** [comment]
> Is it too late to add an overload of MULX that only returns the high half (ulong MultiplyNoFlagsHigh(ulong left, ulong right))?

Somewhat ironically it's easier to implement the version that returns only the high half than the actual MULX. And it's also likely to be more useful.

**Comment by fiigii** [comment]
> looks like we still have 3 APIs that aren't implemented yet.

@tannergooding Yes, I will implement them in my 64-bit intrinsic PR.

**Comment by fiigii** [comment]
@mikedn @pentp It is a good idea to provide a "safe" version of MLUX intrinsic that does not touch memory pointers. So, here we may have four overload options
1. return low and out arg for high (we have this now)
2. return high and out arg for low
3. return low only
4. return high only

Do you think we should have them all or just "3"?

**Comment by mikedn** [comment]
> return low only

Who would be using that and for what? Even the current form has dubious value, a form that returns only the low half is practically useless.

**Comment by tannergooding** [comment]
The operation does the following:
> Performs an unsigned multiplication of the implicit source operand (EDX/RDX) and the specified source operand
(the third operand) and stores the low half of the result in the second destination (second operand), the high half
of the result in the first destination operand (first operand), without reading or writing the arithmetic flags. This
enables efficient programming where the software can interleave add with carry operations and multiplications.
>
> If the first and second operand are identical, it will contain the high half of the multiplication result.

So, the instruction supports returning both the high and low or just the high. There is not an actual way to return just the low (you have to allocate a register, even if it is just thrown away).

**Comment by mikedn** [comment]
> The operation does the following:

Yeah. We know that.

> So, the instruction supports returning both the high and low or just the high. There is not an actual way to return just the low (you have to allocate a register, even if it is just thrown away).

Trouble is, you don't know what the problem is with implementing the proposed form.

**Comment by tannergooding** [comment]
> Trouble is, you don't know what the problem is with implementing the proposed form.

I was commenting in regards to the overloads of the `MULX` instruction that could be exposed. If we were breaking the available functionality apart, expose one that returns both and one that returns just upper makes sense based on how the instruction operates.

As for handling the instruction, I would imagine it to be similar to (but not quite the same as) `div` or `mul`, where we have cases with implicit sources (some overloads of `mul` implicitly take EDX:EAX) and where we have cases where multiple destination registers are written (all overloads of DIV write to both EDX:EAX for the result). There is likely some additional complexity that also needs to be handled, but I'm sure that it can be resolved. 😄

**Comment by fiigii** [comment]
> Being able to multiply two 64 bit values and obtain a 128 bit result is useful (though not necessarily as MULX).

Right, most of the MULX scenarios seem to need low and high results both (large int computation, large int conversion, etc.).

>  expose one that returns both and one that returns just upper makes sense based on how the instruction operates.

My question is do we really need "high half only" in some workloads? If not very common, perhaps we can change the API to "return high and out arg for low", which may be better for JIT implementation and optimization.

Additionally, is it possible to expose this kind of APIs via multi-return-value, like `Tuple<ulong, ulong> MultiplyNoFlagsHigh(ulong left, ulong right)`?

**Comment by tannergooding** [comment]
> My question is do we really need "high half only" in some workload

At least as of right now, we have only reviewed and approved the API that returns both. It might be worth waiting to do this one until after we can review it again and determine the appropriate semantics to expose (which should be the return and which should be the out, and if we want to expose an overload that only returns upper).

> Additionally, is it possible to expose this kind of APIs via multi-return-value, like Tuple<ulong, ulong> MultiplyNoFlags(ulong left, ulong right)

The current API guidelines say we avoid using tuples in public APIs.

**Comment by tannergooding** [comment]
> If not very common, perhaps we can change the API to "return high and out arg for low", which may be better for JIT implementation and optimization.

I think this is the version I would prefer to go to API review, as it makes it easier to expose the overload that only returns `upper` if we got enough feedback requesting it.

**Comment by fiigii** [comment]
> I think this is the version I would prefer to go to API review, as it makes it easier to expose the overload that only returns upper if we got enough feedback requesting it.

Agree. Should other people give a permission before I make the API change?

**Comment by tannergooding** [comment]
> Agree. Should other people give a permission before I make the API change?

For now, we should skip this just this method and pull out the proposed changes into a new issue. We might be able to get to it in the next triage, but that depends on a number of things (such as the other issues in the queue, if we have a dedicated session scheduled instead, etc).

We should still be able to implement `BitFieldExtract` and `ZeroHighBits` though.

**Comment by fiigii** [comment]
> We should still be able to implement BitFieldExtract and ZeroHighBits though.

Got it, will do.

**Comment by tannergooding** [comment]
For the issue covering `MultiplyNoFlags`, I think we have two options (talked briefly with @GrabYourPitchforks about this, since he is also on the API review committee):

```csharp
public static unsafe uint MultiplyHighNoFlags(uint left, uint right);
public static unsafe uint MultiplyNoFlags(uint left, uint right, uint* high);

// or

public static unsafe uint MultiplyNoFlags(uint left, uint right);
public static unsafe uint MultiplyNoFlags(uint left, uint right, uint* lower);
```

The former matches the current native API shape (`_mulx_u32` doesn't have an equivalent that only returns the high bits, but there is `__mul128` and `__mulh`, for comparison).
The latter likely better matches the actual instruction "shape", but may cause issues if users are translating code from an existing native implementation

**Comment by pentp** [comment]
The `MultiplyNoFlags` overload that returns both parts as currently proposed is fine and should be implemented anyway (it's useful for 64 * 64 -> 128bit multiplication).
The overload that returns only the high part is useful when doing 64-bit division by multiplication with a pre-computed reciprocal from a table.

Both would be really helpful in `System.Decimal`.

**Comment by fiigii** [comment]
> public static unsafe uint MultiplyNoFlags(uint left, uint right);
public static unsafe uint MultiplyNoFlags(uint left, uint right, uint* lower);

For this version, can we combine them to one API via optional argument?
```csharp
public static unsafe uint MultiplyNoFlags(uint left, uint right, uint* lower = null);
```

**Comment by pentp** [comment]
I think two separate overloads is more in line with the existing APIs.

**Comment by tannergooding** [comment]
The general consensus is to avoid optional parameters in public APIs.

**Comment by fiigii** [comment]
Ok, just curious. 

I prefer the second one (matches the actual instruction "shape"), and thank @pentp for providing the real scenario 😄
